src/boot/efi/boot.c

```diff
@@ -1951,6 +1961,95 @@ static VOID config_entry_add_linux(
+#endif
+static const EFI_GUID fdt_guid = DEVICE_TREE_GUID;
+
+static EFI_STATUS devicetree_get(const void **fdt) {
+        EFI_CONFIGURATION_TABLE *entry = ST->ConfigurationTable;
+        EFI_CONFIGURATION_TABLE *end = entry + ST->NumberOfTableEntries;
+
+        for (; entry < end; entry++) {
+                if (CompareGuid(&entry->VendorGuid, (EFI_GUID *)&fdt_guid) == 0) {
+                        *fdt = entry->VendorTable;
+                        return EFI_SUCCESS;
+                }
+        }
+
+        return EFI_UNSUPPORTED;
+}
+
+static const void *devicetree_orig;
+static EFI_PHYSICAL_ADDRESS devicetree_addr;
+static UINTN devicetree_pages;
+
+static EFI_STATUS devicetree_install(EFI_FILE_HANDLE root_dir, const CHAR16 *name) {
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE handle = NULL;
+        _cleanup_freepool_ EFI_FILE_INFO *info = NULL;
+        UINTN size;
+        EFI_STATUS err;
+        void *fdt;
+
+        err = devicetree_get(&devicetree_orig);
+        if (EFI_ERROR(err))
+                return err;
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &handle, (CHAR16 *)name,
+                        EFI_FILE_MODE_READ, EFI_FILE_READ_ONLY);
+        if (EFI_ERROR(err))
+                return err;
+
+        info = LibFileInfo(handle);
+        if (!info)
+                return EFI_OUT_OF_RESOURCES;
+        if (info->FileSize < FDT_V1_SIZE || info->FileSize > 32 * 1024 * 1024)
+                /* 32MB device tree blob doesn't seem right */
+                return EFI_INVALID_PARAMETER;
+
+        size  = info->FileSize;
+        FreePool(TAKE_PTR(info));
+
+        devicetree_pages = (size + EFI_PAGE_SIZE - 1) / EFI_PAGE_SIZE;
+        err = uefi_call_wrapper(BS->AllocatePages, 4, AllocateAnyPages, EfiACPIReclaimMemory,
+                        devicetree_pages, &devicetree_addr);
+        if (EFI_ERROR(err))
+                return err;
+
+        fdt = (void *)(uintptr_t)devicetree_addr;
+
+        err = uefi_call_wrapper(handle->Read, 3, handle, &size, fdt);
+        if (EFI_ERROR(err))
+                goto err_free;
+
+        if (fdt_check_full(fdt, size)) {
+                err = EFI_INVALID_PARAMETER;
+                goto err_free;
+        }
+
+        err = uefi_call_wrapper(BS->InstallConfigurationTable, 2,
+                        (EFI_GUID *)&fdt_guid, (VOID *)fdt);
+        if (EFI_ERROR(err))
+                goto err_free;
+
+        return err;
+err_free:
+        uefi_call_wrapper(BS->FreePages, 2, devicetree_addr, devicetree_pages);
+        return err;
+}
+
+static EFI_STATUS devicetree_restore(void) {
+        EFI_STATUS err;
+
+        err = uefi_call_wrapper(BS->InstallConfigurationTable, 2,
+                        (EFI_GUID *)&fdt_guid, (VOID *)devicetree_orig);
+        if (EFI_ERROR(err))
+                return err;
+
+        return uefi_call_wrapper(BS->FreePages, 2, devicetree_addr, devicetree_pages);
+}
```

Comment:
plese move these functions into a new .c file, boot.c is already too larger, and this is kinda non-core stuff, so let's split this out

Reply:
Will do